src/metabase/lib/stage.cljc

```diff
@@ -354,6 +354,13 @@
                                               (lib.util/query-stage query previous-stage-number)
                                               style)))))
 
+(mu/defn is-stage-empty :- :boolean
+  "Is the given query stage empty of clauses?"
+  [query        :- ::lib.schema/query
+   stage-number :- :int]
+  (every? #(nil? (get (lib.util/query-stage query stage-number) %))
+          [:expressions :filters :limit :breakout :aggregation :joins :order-by :fields]))
```

Comment:
> I wonder if something like would be simpler and more robust in the face of change.

Not sure if this applies to Clojure world, but the reason I listed all the clauses here comes from my experience with [`Pick`](https://www.typescriptlang.org/docs/handbook/utility-types.html#picktype-keys) and [`Omit`](https://www.typescriptlang.org/docs/handbook/utility-types.html#omittype-keys) in TypeScript - it's much more difficult to maintain types defined with `Omit` over time, as it's easier for properties to "slip" into other types unnoticed.
Same goes for [`_.pick`](https://underscorejs.org/#pick) and [`_.omit`](https://underscorejs.org/#omit) in underscore/lodash/etc.

Please let me know if you believe it should be defined in omit-like fashion.

> I also wonder if it's a good idea for this function to return `true` for the first stage if it has only a source table/card.

FE relies on this. If BE thinks otherwise, then perhaps this function should be called differently. 
Originally it was defined in FE as an inverse and was called `hasAnyClauses`.

Reply:
I prefer the "omit" way. (We don't have the problem with types you describe.)

> perhaps this function should be called differently

I find `is-stage-empty` misleading. `has-clauses?` (for the complementary logic) less so, because I can squint in a way that doesn't make the source table a clause. I had `trivial-stage?` in mind, but 'm not satisfied with that either, because normally you wouldn't know what "trivial" means for a stage.

Reply:
> I prefer the "omit" way. (We don't have the problem with types you describe.)

Updated: 60062b3eebd6ccc3d1a84ae2290e5792cefcf5f9

> I find `is-stage-empty` misleading. `has-clauses?` (for the complementary logic) less so, because I can squint in a way that doesn't make the source table a clause. I had `trivial-stage?` in mind, but 'm not satisfied with that either, because normally you wouldn't know what "trivial" means for a stage.

Inversed and renamed to `has-clauses`: 6d246b1235f3f35dc5b78aa6ff7d248d2000cb0f